### PR TITLE
feature/import

### PR DESCRIPTION
**Pull request by subdavis**
This is **not complete**.  I wanted to go ahead and start a discussion to talk about my approach.

It looks like I'll need to rebase so my iOS fixes aren't a part of this PR.

### CODE REVIEW

File: src/import.js [hunk 1]
```diff
@@ -1,284 +1,89 @@
...
-          console.log('Retrying ' + failedPromises + ' calls');
-        }
+  const body = utils.normalizeBody(JSON.parse(event.body));
 
-        callback(null, output);
-      });
-  } catch (e) {
-    console.error('Exception processing import data', e);
-    callback(null, 'ERROR: ' + e);
+  if (typeof (body.ciphers) !== 'object') {
```
Comment (vvondra) [inline src/import.js]: `Array.isArray` ?

File: src/import.js [hunk 2]
```diff
@@ -1,284 +1,89 @@
...
-    user = await loadUser(event);
-  } catch (e) {
-    callback(null, e.message);
-    return;
+  for (let i = 0; i < body.ciphers.length; i += 1) {
+    const rawCipher = body.ciphers[i];
+    savePromises.push(Cipher.createAsync(buildCipherDocument(rawCipher, user))
+      .then(result => ({ success: true, result, rawCipher }))
+      .catch(error => ({ success: false, error, rawCipher })));
   }
```
Comment (vvondra) [inline src/import.js]: a little more idiomatic JS is
```js
body.ciphers.forEach((rawCipher) => {
    savePromises.push(Cipher.createAsync(buildCipherDocument(rawCipher, user))
      .then(result => ({ success: true, result, rawCipher }))
      .catch(error => ({ success: false, error, rawCipher })));
});

**Comment by vvondra** [comment]
I'm looking forward to this one, feel free to remove the other instructions in the README.md and remove all the obsolete references not using the CLI.

**Comment by subdavis** [comment]
Yo @vvondra - what is `createAsync` - I've looked all through the dynogels library and it's nowhere to be found in the code or docs.  I found reference in a few issues, but there's no documentation on what comes out of the returned promise.  I also checked `aws-sdk` and `createAsync` isn't mentioned.  What library provides this function??

**Comment by vvondra** [comment]
@subdavis its this https://www.npmjs.com/package/dynogels-promisified

dynogels basic API only gives you callback style method calls, that get's quickly quite messy.

**Comment by subdavis** [comment]
@vvondra I believe this PR is now complete.  Please review and let me know where changes are needed.  Thank you.

**Comment by vvondra** [comment]
there's a problem with the provisioned dynamodb throughput

```
2018-08-13 09:40:56.161 (+02:00)	1b18d856-9ecc-11e8-9567-6d2e49fd95ed	{ err: 
   { ProvisionedThroughputExceededException: The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API.
```

to make this project cheap to run the default throughput on the dynamodb tables is quite low and it's not enough for the import.

i worked around it before with the ugly while loops doing the retry until all the promises succeed, but now since we call it through API gateway (before it was a direct lambda function invoke) we have a hard limit of 30s.

i see two options to make this function not be frustrating for people:
a) in the import function modify the table to increase throughput temporarily through the AWS SDK (ill need to add UpdateTable IAM permissions to the import 
function)
b) deploy a queuing setup through SQS. the import function will just write the import jobs to SQS via sqs:SendMessageBatch and another import function will process the messages and update one by one

the first seems a bit simpler, ill try to add it and remove the crazy while loops clearing the list of promises

**Comment by subdavis** [comment]
I agree - I'll give 'a' a try.

What sort of import size are you doing to cause throughout errors?  I would like to reproduce this myself.

**Comment by vvondra** [comment]
i exported my old lastpass account with about 600 passwords and 20 folders, the lastpass csv import format, nothing fancy

ping me if you need a hand, just push what you have WIP

**Comment by vvondra** [comment]
I see. You're suggestion sounds reasonable. Let's just update the capacity there and we're good to merge.

**Comment by subdavis** [comment]
This ought to do it.  LMK if there's anything else.

**Comment by subdavis** [comment]
Anything else that needs to be done here, @vvondra ?

**Comment by vvondra** [comment]
Thanks for this one!

**Pull request merged**